motionplan/ik/nloptInverseKinematics.go

```diff
@@ -165,32 +178,52 @@ func (ik *NloptIK) Solve(ctx context.Context,
 		}
 	}
 
-	select {
-	case <-ctx.Done():
-		ik.logger.Info("solver halted before solving start; possibly solving twice in a row?")
-		return err
-	default:
-	}
-
+	solveChan := make(chan *optimizeReturn, 1)
+	defer close(solveChan)
 	for iterations < ik.maxIterations {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
 		default:
 		}
+
+		var solutionRaw []float64
+		var result float64
+		var nloptErr error
+
 		iterations++
-		solutionRaw, result, nloptErr := opt.Optimize(referenceframe.InputsToFloats(startingPos))
```

Comment:
Can you motivate why this is changing?

Reply:
This was causing tests to flake out. Surprised we haven't seen more of it more recently.
There's nothing internal to the actual `nlopt` call that takes a context. We need to have something watching the `ctx` and calling `opt.ForceStop()`, in parallel with the actual `Optimize` call. Otherwise this can cause outer callers to exit while the inner `Optimize` is still running, orphaning goroutines.

Reply:
Sounds like a good improvement thanks for making it!